src/ray/core_worker/core_worker.cc

```diff
@@ -867,7 +867,11 @@ Status CoreWorker::Put(const RayObject &object,
   reference_counter_->AddOwnedObject(
       *object_id, contained_object_ids, rpc_address_, CurrentCallSite(), object.GetSize(),
       /*is_reconstructable=*/false, NodeID::FromBinary(rpc_address_.raylet_id()));
-  return Put(object, contained_object_ids, *object_id, /*pin_object=*/true);
+  auto status = Put(object, contained_object_ids, *object_id, /*pin_object=*/true);
+  if (!status.ok()) {
+    reference_counter_->RemoveOwnedObject(*object_id);
+  }
+  return status;
```

Comment:
As a driveby, in `CoreWorker::Put`, if the Plasma creation fails, we delete the owned object reference.

Reply:
I think we discussed this offline, but what's the reason that we can't just add the reference *after* putting it in plasma? Is it because putting it in plasma triggers the pinning protocol which will make the `WaitForObjOutOfScope` RPC to this worker? It seems like we should be able to address this by reordering so that we (1) put object in plasma, (2) check if it succeeded, (3) if succeded, add to ref counter and send pin RPC.

That way, we could avoid adding this `RemoveOwnedObject` method entirely. Not a big deal though if that's a messy change, this case shouldn't happen under normal operation.

Reply:
I think that would solve the pin RPC ordering, but that wouldn't solve the ownership-based object directory issue that @zhuohan123 was running into for the `Create()` + `Seal()` path in which the reference is added after the Plasma put. In that case, the Plasma put triggers an [`AddObjectLocationOwner` RPC](https://github.com/ray-project/ray/blob/97211a6170f656fca5f09122646ba37da6413871/src/ray/object_manager/ownership_based_object_directory.cc#L82-L89) to the worker that can race against the owned reference being added to the reference table.

Given that this happens at a much lower level, hooked in as a Plasma store notification callback, I don't think that this protocol would be as easily tweaked.

Reply:
Right, makes sense, what we currently have LGTM then.